kernel/uint63_31.ml

```diff
@@ -118,27 +118,29 @@ let div21 xh xl y =
 let div21 xh xl y =
   if Int64.compare y xh <= 0 then zero, zero else div21 xh xl y
 
-     (* exact multiplication *)
+(* exact multiplication *)
 let mulc x y =
-  let lx = ref (Int64.logand x maxuint31) in
-  let ly = ref (Int64.logand y maxuint31) in
+  let lx = Int64.logand x maxuint31 in
+  let ly = Int64.logand y maxuint31 in
   let hx = Int64.shift_right x 31 in
   let hy = Int64.shift_right y 31 in
-  let hr = ref (Int64.mul hx hy) in
-  let lr = ref (Int64.logor (Int64.mul !lx !ly) (Int64.shift_left !hr 62)) in
-  hr := (Int64.shift_right_logical !hr 1);
-  lx := Int64.mul !lx hy;
-  ly := Int64.mul hx !ly;
-  hr := Int64.logor !hr (Int64.add (Int64.shift_right !lx 32) (Int64.shift_right !ly 32));
-  lr := Int64.add !lr (Int64.shift_left !lx 31);
-  hr := Int64.add !hr (Int64.shift_right_logical !lr 63);
-  lr := Int64.add (Int64.shift_left !ly 31) (mask63 !lr);
-  hr := Int64.add !hr (Int64.shift_right_logical !lr 63);
-  if Int64.logand !lr Int64.min_int <> 0L
-  then Int64.(sub !hr one, mask63 !lr)
-  else (!hr, !lr)
-
-let equal x y = mask63 x = mask63 y
+  (* compute the median products *)
+  let s = Int64.add (Int64.mul lx hy) (Int64.mul hx ly) in
+  (* s fits on 64 bits, split it into a 33-bit high part and a 31-bit low part *)
+  let lr = Int64.shift_left (Int64.logand s maxuint31) 31 in
+  let hr = Int64.shift_right_logical s 31 in
+  (* add the outer products *)
+  let lr = Int64.add (Int64.mul lx ly) lr in
+  let hr = Int64.add (Int64.mul hx hy) hr in
```

Comment:
Can't hr overflow by one to 2^64? As far as I understand hr can be up to
(2^32-1)*(2^32-1)+(2^33-1) = 2^64 - 2^33 + 1 + 2^33 -1 = 2^64
Is this taken care of below?

Reply:
Ah no, it is fine since the high part of s cannot be 2^33-1:
The maximum value of s is
2*(2^31-1)(2^32-1)=2*(2^63-2^31-2^32+1)=2^64-2^32-2^33+2
So if this is right shifted by 31, we get 2^33-2^1-2^2 = 2^33-6.
Since this is not so obvious (at least to me), you might want to document this.

Reply:
Your reasoning is right, but there is a much simpler way to prove the absence of overflow: We are computing a 126-bit product and `hr` contains the part of the result starting at bit 62.

Reply:
Indeed, I was a bit in a VST style C verification forward analysis mode. I think it still doesn't hurt to document things if a simple forward bit counting analysis is not sufficient to guarantee results are in range.

Pondering over my "L suffix means they introduced a 63 bit integer type in 32 bit OCaml" confusion: in hindsight it is clear. Only if others reading this say they haven't seen this "L" suffix syntax before, I would think about putting in a short hint that this is the syntax for 64 bit integers. I still think that not that many people think about 63/64 bit integer arithmetic on 32 bit platforms so knowledge on the details might be less widespread than an expert in this area like you might think.